File: solutions/20 Valid Parantheses.py
#Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

#An input string is valid if:

 #   Open brackets must be closed by the same type of brackets.
 #   Open brackets must be closed in the correct order.
 #   Every close bracket has a corresponding open bracket of the same type.






class Solution:
    def isValid(self, s: str) -> bool:
        
        opening_brackets = "([{"
        closing_brackets = ")]}"

        stack = []
        for char in s:
            if char in opening_brackets:
                stack.append(char)
            elif char in closing_brackets:
                if not stack or stack.pop() != opening_brackets[closing_brackets.index(char)]:
                    return False
        

        return len(stack) == 0

File: solutions/20-Valid-Parantheses.java
import java.util.Stack;

class Solution {

public static boolean isValid(String s) {		
    Stack<Character> newStack = new Stack();

	if(s.length() == 1 || s.charAt(0) == ')' ||s.charAt(0) == ']' || s.charAt(0) == '}') {
		return false;
	}


		for (int i = 0; i < s.length(); i++) {
			if(s.charAt(i) == '(' ||s.charAt(i) == '[' || s.charAt(i) == '{') {
				newStack.push(s.charAt(i));
				System.out.println(s.charAt(i));
			} else {
                if(newStack.isEmpty()) {
					return false;
				}
				if(s.charAt(i) == ')' && newStack.pop() != '(') {
					return false;
				} else if(s.charAt(i) == ']' && newStack.pop() != '[') {
					return false;
				} else if(s.charAt(i) == '}' && newStack.pop() != '{') {
					return false;
				}
			}
			
			
		}
		
		if(newStack.isEmpty()) {
			return true;
		} else {
			return false;
		}
	    	
	    	
	    	
	    	
	}
}
